File: problem/Min_Distance_Word_Fingers/minimumDistance.java
class Solution {
    private int cal(int a, int b) {
        return Math.abs(a / 6 - b / 6) + Math.abs(a % 6 - b % 6);
    }

    public int minimumDistance(String word) {
        int n = word.length();
        int[] dp = new int[26];

        for (int i = 1; i < n; i++) {
            int[] ndp = new int[26];
            int p = word.charAt(i - 1) - 'A';
            int t = word.charAt(i) - 'A';

            for (int j = 0; j < 26; j++) {
                ndp[j] = dp[j] + cal(p, t);
            }
            
            ndp[p] = Integer.MAX_VALUE;
            for (int j = 0; j < 26; j++) {
                ndp[p] = Math.min(ndp[p], dp[j] + cal(j, t));
            }
            dp = ndp;
        }

        int res = dp[0];
        for (int val : dp) res = Math.min(res, val);
        return res;
    }
}

File: problem/Min_Distance_Word_Fingers/minimumDistance.py
class Solution:
    def minimumDistance(self, word):
        def cal(a, b):
            return abs(a // 6 - b // 6) + abs(a % 6 - b % 6)

        n = len(word)
        dp = [0] * 26
        
        for i in range(1, n):
            p = ord(word[i-1]) - ord('A')
            t = ord(word[i]) - ord('A')
            
            ndp = [d + cal(p, t) for d in dp]
            
            min_val = float('inf')
            for j in range(26):
                min_val = min(min_val, dp[j] + cal(j, t))
            
            ndp[p] = min(ndp[p], min_val)
            dp = ndp
            
        return min(dp)